Describe local test run

There are no descriptions or prepared scripts (like make-commands) for test run. Setting up test environment requires using mage -- it is not a default `go mod tidy; go test .` run and not obvious for one unfamiliar with mage. Running a sequence of commands from CI was not enough for me: it failed with
```bash
Running unit tests...
../generate_templates_vfs.go:9:2: no required module provides package github.com/shurcooL/vfsgen; to add it:
	go get github.com/shurcooL/vfsgen
cli/codegen/static/static.go:5: running "go": exit status 1
Error: running "go generate -tags=dev ./cli/codegen/static" failed with exit code 1
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -44,6 +44,7 @@ and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.
 - Tarantool benchmark tool (early alpha, API can be changed in the near future).
 - Ability to reverse search in ``cartridge enter`` and ``cartridge connect`` commands.
 - Added support for functionality from Golang 1.17.
+- Describe scenario of local test run.
 
 ## [2.10.0] - 2021-07-28
 

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -36,6 +36,7 @@ and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.
 - Fixed project build with capital letters in the project name.
 - Fixed display of Docker image pull (``cartridge pack`` command with ``--verbose`` flag).
 - Added support for new Tarantool release policy.
+- ``mage clean`` removes all generated code.
 
 ### Added
 

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -37,6 +37,7 @@ and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.
 - Fixed display of Docker image pull (``cartridge pack`` command with ``--verbose`` flag).
 - Added support for new Tarantool release policy.
 - ``mage clean`` removes all generated code.
+- All mage test commands now depend on code generation step.
 
 ### Added
 

--- a/README.dev.md
+++ b/README.dev.md
@@ -1,0 +1,95 @@
+## How to run tests
+
+### Requirements
+
+1. Install [Go](https://go.dev/doc/install) 1.17.
+   ```bash
+   go version
+   ```
+
+2. Install [Python](https://www.python.org/downloads/) 3.x and [pip](https://pypi.org/project/pip/).
+   ```bash
+   python3 --version
+   pip3 --version
+   ```
+
+3. Install `unzip`, `rpm` and `cpio` packages.
+   ```bash
+   unzip -v
+   rpm --version
+   cpio --version
+   ```
+   
+4. Install [git](https://git-scm.com/downloads).
+   ```bash
+   git --version
+   ```
+      
+5. Install [docker](https://www.docker.com/get-started).
+   ```bash
+   docker --version
+   ```
+    
+6. Install [Tarantool](https://www.tarantool.io/en/download/os-installation/) (1.10 or 2.x version).
+   ```bash
+   tarantool --version
+   ```
+
+7. Install [mage](https://github.com/magefile/mage).
+   ```bash
+   mage --version
+   ```
+   If something went wrong, this may help.
+   ```bash
+   export PATH=$(go env GOPATH)/bin:$PATH
+   ```
+     
+8. Install [pytest](https://docs.pytest.org/en/6.2.x/getting-started.html).
+   ```bash
+   python3 -m pytest --version
+   ```
+
+9. Clone this repo.
+   ```bash
+   git clone git@github.com:tarantool/cartridge-cli.git
+   cd ./cartridge-cli
+   ```
+   
+10. To run tests, git user must be configured. For example,
+    ```bash
+    git config --global user.email "test@tarantool.io"
+    git config --global user.name "Tar Antool"
+    ```
+   
+11. Install pytest dependencies.
+    ```bash
+    pip3 install -r test/requirements.txt
+    ```
+    
+12. Install luacheck.
+    ```bash
+    tarantoolctl rocks install luacheck
+    ```
+
+All remaining dependencies (like code generation) will be invoked with mage if needed.
+
+### Test run
+
+To run all tests, call
+```bash
+mage test
+```
+
+You can run specific test sections.
+```bash
+# Static code analysis (including tests code).
+mage lint
+# Go unit tests.
+mage unit
+# pytest integration tests.
+mage integration
+# Run test example with pytest.
+mage testExamples
+# pytest end-to-end tests for packages.
+mage e2e
+```

--- a/magefile.go
+++ b/magefile.go
@@ -28,7 +28,7 @@ var packagePath = "./cli"
 var generateModePath = filepath.Join(packagePath, "codegen", "generate_code.go")
 var generatedFilesPath = fmt.Sprintf("./%s", filepath.Join(packagePath, "codegen", "static"))
 
-var generatedFSFile = "cartridge_vfsdata_gen.go"
+var generatedFSFile = "../../create/create_vfsdata_gen.go"
 var generatedModeFile = "create_cartrdige_template_filemodes_gen.go"
 
 var completionPath = "./completion"

--- a/magefile.go
+++ b/magefile.go
@@ -96,9 +96,35 @@ func init() {
 	os.Setenv("GO111MODULE", "on")
 }
 
+// Generate Go code that statically implements filesystem
+// and map with modes for that filesystem.
+func GenerateGoCode() error {
+	fmt.Println("Generating Go code...")
+
+	err := sh.RunWith(
+		getBuildEnv(), goExe,
+		"generate", "-tags=dev",
+		generatedFilesPath,
+	)
+
+	if err != nil {
+		return err
+	}
+
+	err = sh.RunWith(
+		getBuildEnv(), goExe,
+		"run", generateModePath,
+	)
+
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
 // Run go vet and flake8
 func Lint() error {
-	fmt.Println("Generating Go code...")
 	mg.Deps(GenerateGoCode)
 
 	fmt.Println("Running go vet...")
@@ -121,8 +147,9 @@ func Lint() error {
 
 // Run unit tests
 func Unit() error {
-	fmt.Println("Running unit tests...")
 	mg.Deps(GenerateGoCode)
+
+	fmt.Println("Running unit tests...")
 
 	if mg.Verbose() {
 		return sh.RunV(goExe, "test", "-v", "./cli/...")
@@ -133,18 +160,24 @@ func Unit() error {
 
 // Run integration tests
 func Integration() error {
+	mg.Deps(GenerateGoCode)
+
 	fmt.Println("Running integration tests...")
 	return sh.RunV(py3Exe, "-m", "pytest", "test/integration")
 }
 
 // Run examples tests
 func TestExamples() error {
+	mg.Deps(GenerateGoCode)
+
 	fmt.Println("Running examples tests...")
 	return sh.RunV(py3Exe, "-m", "pytest", "test/examples")
 }
 
 // Run e2e tests
 func E2e() error {
+	mg.Deps(GenerateGoCode)
+
 	fmt.Println("Running e2e tests...")
 	return sh.RunV(py3Exe, "-m", "pytest", "test/e2e")
 }
@@ -156,11 +189,11 @@ func Test() {
 
 // Build cartridge-cli executable
 func Build() error {
+	mg.Deps(GenerateGoCode)
+
 	var err error
 
 	fmt.Println("Building...")
-
-	mg.Deps(GenerateGoCode)
 
 	err = sh.RunWith(
 		getBuildEnv(), goExe, "build",
@@ -173,31 +206,6 @@ func Build() error {
 
 	if err != nil {
 		return fmt.Errorf("Failed to build cartridge-cli executable: %s", err)
-	}
-
-	return nil
-}
-
-// Generate Go code that statically implements filesystem
-// and map with modes for that filesystem.
-func GenerateGoCode() error {
-	err := sh.RunWith(
-		getBuildEnv(), goExe,
-		"generate", "-tags=dev",
-		generatedFilesPath,
-	)
-
-	if err != nil {
-		return err
-	}
-
-	err = sh.RunWith(
-		getBuildEnv(), goExe,
-		"run", generateModePath,
-	)
-
-	if err != nil {
-		return err
 	}
 
 	return nil

